src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => {
+        delete newState[key];
+      });
+    } else if (action.type === 'clear') {
+      newState = {};
+    }
+    history.push({ ...newState });
+    currentState = { ...newState };
```

Comment:
There is no need to create a new copy of `newState` when assigning it to `currentState`. You can simply assign `currentState = newState;` since `newState` is already a fresh copy.